halo2_proofs/src/dev.rs

```diff
@@ -97,16 +97,120 @@ impl Region {
     /// An unassigned cell.
     Unassigned,
     /// A cell that has been assigned a value.
     Assigned(F),
+    /// A value stored as a fraction to enable batch inversion.
+    Rational(F, F),
     /// A unique poisoned cell.
     Poison(usize),
 }
 
+impl<F: Group + Field> PartialEq for CellValue<F> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Unassigned, Self::Unassigned) => true,
+            (Self::Assigned(a), Self::Assigned(b)) => a == b,
+            (Self::Rational(a, b), Self::Rational(c, d)) => *a * d == *b * c,
+            (Self::Assigned(a), Self::Rational(n, d)) => *a == *n * d.invert().unwrap(),
+            (Self::Rational(n, d), Self::Assigned(a)) => *n * d.invert().unwrap() == *a,
+            (Self::Poison(a), Self::Poison(b)) => a == b,
+            _ => false,
+        }
+    }
+}
+
+impl<F: Group + Field> CellValue<F> {
+    /// Returns the numerator.
+    pub fn numerator(&self) -> Option<F> {
+        match self {
+            Self::Rational(numerator, _) => Some(*numerator),
+            _ => None,
+        }
+    }
+
+    /// Returns the denominator
+    pub fn denominator(&self) -> Option<F> {
+        match self {
+            Self::Rational(_, denominator) => Some(*denominator),
+            _ => None,
+        }
+    }
+}
+
+impl<F: Group + Field> From<Assigned<F>> for CellValue<F> {
+    fn from(value: Assigned<F>) -> Self {
+        match value {
+            Assigned::Zero => CellValue::Unassigned,
+            Assigned::Trivial(value) => CellValue::Assigned(value),
+            Assigned::Rational(numerator, denominator) => {
+                CellValue::Rational(numerator, denominator)
+            }
+        }
+    }
+}
+
+fn calculate_assigned_values<F: Group + Field>(
+    cell_values: &mut [CellValue<F>],
+    inv_denoms: &[Option<F>],
+) {
+    assert_eq!(inv_denoms.len(), cell_values.len());
+    for (value, inv_den) in cell_values.iter_mut().zip(inv_denoms.iter()) {
+        // if numerator and denominator exist, calculate the assigned value
+        // otherwise, return the original CellValue
+        *value = match value {
+            CellValue::Rational(numerator, _) => CellValue::Assigned(*numerator * inv_den.unwrap()),
+            _ => *value,
+        };
+    }
+}
+
+fn batch_invert_cellvalues<F: Field + Group>(cell_values: &mut [Vec<CellValue<F>>]) {
+    let mut denominators: Vec<_> = cell_values
+        .iter()
+        .map(|f| {
+            f.par_iter()
+                .map(|value| value.denominator())
+                .collect::<Vec<_>>()
+        })
+        .collect();
+
+    let mut mut_denominators = denominators
+        .iter_mut()
+        .flat_map(|f| {
+            f.iter_mut()
+                // If the denominator is trivial, we can skip it, reducing the
+                // size of the batch inversion.
+                .filter_map(|d| d.as_mut())
+        })
+        .collect::<Vec<_>>();
```

Comment:
We can construct a 2-d vec here s.t. we do not need to use `split_off` to split large vec into chunks.